File: src/main/java/org/jboss/netty/example/http/websocketx/server/Room.java
package org.jboss.netty.example.http.websocketx.server;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.jboss.netty.channel.ChannelHandlerContext;
import org.jboss.netty.handler.codec.http.websocketx.TextWebSocketFrame;

import com.google.gson.annotations.Expose;

public class Room implements Serializable {

	@Expose
	private String id;

	@Expose
	private boolean isAdult;

	@Expose
	private String name;
	
	@Expose
	private int playerCount;
	
	private Map<String,Player> players = new HashMap<String,Player>();
	
	private List<Round> rounds = new ArrayList<Round>(5);
	
	private Round currentRound;
	
	private State state = State.CHATTING;
	
	public State getState() {
		return state;
	}
	
	public enum State {
		CHATTING,VOTING,WRITING_ACRONYMS,FACE_OFF
	}

	public String getId() {
		return id;
	}
	
	public String getName() {
		return name;
	}

	public int getPlayerCount() {
		return playerCount;
	}
	
	public Collection<Player> getPlayers() {
		return players.values();
	};
	
	public List<Player> getLeaders() {
		List<Player> leaders = new ArrayList<Player>(players.values());
		Collections.sort(leaders, new Comparator<Player>() {
		
			public int compare(Player a, Player b) {
				int c = Integer.valueOf(a.getTotalVoteCount()).compareTo(b.getTotalVoteCount());
				if(c!=0) {
					return c;
				}
				if(rounds.isEmpty()) {
					return c;
				}
				Round r = rounds.get(rounds.size()-1);
				Acronym aa = r.getAnswer(a.getUserId());
				Acronym bb = r.getAnswer(b.getUserId());
				if(aa==null) {
					return 1;
				}
				if(bb==null) {
					return -1;
				}
				return Long.valueOf(aa.getReceived()).compareTo(bb.getReceived());
			}
			
		});
		return leaders;
	}

	public int getRoomSize() {
		return players.size();
	}
	
	public boolean hasEnoughPlayers() {
		return players.size() > 2;
	}
	
	public boolean isAdult() {
		return isAdult;
	}
	
	public boolean isFull() {
		return players.size() > 13;
	}
	
	public void join(ChannelHandlerContext ctx, Request request) {
		String userId = request.getUserId();
		Player player = null;
		if(userId==null) {
			userId = UUID.randomUUID().toString();			
		}
		player = players.get(userId);
		if(player==null) {
			player = new Player(ctx);
			player.setUserId(userId);
			player.setUsername(request.getUsername());
			player.setAvatarUrl(request.getAvatarUrl());
		}
		players.put(userId, player);
		playerCount = players.size();
	}

	public void removePlayer(String userId) {
		players.remove(userId);
		playerCount = players.size();
	}

    public Player getPlayer(String userId) {
        return players.get(userId);
    }

	public void setAdult(boolean isAdult) {
		this.isAdult = isAdult;
	}

	public void setId(String id) {
		this.id = id;
	}

	public void setName(String name) {
		this.name = name;
	}

	public void setPlayerCount(int playerCount) {
		this.playerCount = playerCount;
	}

	public void startRound() {
		state = State.WRITING_ACRONYMS;
	}

	public void startVoting() {
		state = State.VOTING;
	}
	
	public void startChatting() {
		state = State.CHATTING;
	}
	
	public void startFaceOff() {
		state = State.FACE_OFF;
	}
	
}

File: src/main/scala/room.scala
package acro

import java.util.UUID

import scala.collection.JavaConverters._
import java.util.concurrent.TimeUnit
import org.jboss.netty.util.{TimerTask,Timeout}
import org.jboss.netty.example.http.websocketx.server._

import Handler._

object Timer {
  val underlying = new org.jboss.netty.util.HashedWheelTimer
  def apply[U](delay: Long, unit: TimeUnit)(f: => U) {
    underlying.newTimeout(new TimerTask {
      def run(timeout: Timeout) {
        f
      }
    }, delay, unit)
  }
  def seconds[U](delay: Long) = Timer(delay, TimeUnit.SECONDS)_
}
case class Disconnected(userId: String)

class RoomActor(name: String) extends scala.actors.Actor { self =>
  def answerTime = 40
  def voteTime = 20

  val room = new Room()
  room.setName(name)
  room.setAdult(false)
  room.setId(UUID.randomUUID().toString())

  def players = room.getPlayers.asScala

  def cleanup: Unit = {
    for (player <- players)
      if (!player.getContext.getChannel.isConnected)
        self ! Disconnected(player.getUserId)
    Timer.seconds(1)(cleanup)
  }
  Timer.seconds(5)(cleanup)

  var rounds = List.empty[Round]

  def act = loop { react {
    case Join(con) =>
      if (!room.isFull()) {
        room.join(con.channelContext, con.request)
      }
      con.write(gsonHeavy.toJson(new Response("jr", room)))
      val joinedRoom = Handler.gsonHeavy.toJson(new Response("nu",room.getPlayer(con.request.getUserId)))
	  broadcast(joinedRoom)
      if (room.getState == Room.State.CHATTING) {
        startRound()
      }
    case Message(con) =>
      con.request.remove("type")
      con.request.remove("room")
      broadcast(gsonHeavy.toJson(
        new Response("m", con.request.getMessage())))
    case Answer(con) if room.getState == Room.State.WRITING_ACRONYMS =>
      rounds.head.addAnswer(con.request.getUserId,
                            new Acronym(room.getPlayer(con.request.getUserId),
                                        con.request.optString("acronym")))
      val answerCount = Handler.gsonHeavy.toJson(
        new Response("ac", rounds.head.getAcronyms.size))
      broadcast(answerCount)
    case Vote(con) =>
      rounds.head.addVote(con.request.getUserId,
                          con.request.optString("acronym"))
    case Leave(con) =>
      room.removePlayer(con.request.getUserId)
	  val left = Handler.gsonHeavy.toJson(new Response("lv", con.request.getUserId))
      broadcast(left)
    case Disconnected(userId) =>
      println("removing " + userId)
      room.removePlayer(userId)
	  val left = Handler.gsonHeavy.toJson(new Response("lv", userId))
      broadcast(left)
  } }

  def broadcast(str: String) {
    println("broadcasting: " + str)
    for (player <- room.getPlayers.asScala) {
      Handler.write(player.getContext, str)
    }
  }
  val rand = new scala.util.Random

  def startFaceOffRound() {
	  val leaders = room.getLeaders.asScala
  }
  def startRound() {
  val leaders = room.getLeaders.asScala
  if(leaders.head!=null) {
	if(leaders.head.getTotalVoteCount>=30) {
		startFaceOffRound();
	}
  } else
    if (!room.hasEnoughPlayers) {
      room.startChatting()
    } else {

      room.startRound()
      val size = (rounds.size % 4) + 3
      val chars = "ABCDEFGHIJKLMNOPQRSTVW".toSeq
      val acro = rand.shuffle(chars).take(size).mkString
      rounds = new Round :: rounds
		println("\n\new:")
	  for {
	 leader <- room.getLeaders.asScala
	} {
		println(leader.getUsername + " " + leader.getTotalVoteCount)
	}
      rounds.head.setCategory("general")
      rounds.head.setAcronym(acro)
      rounds.head.setRound(rounds.size)
      val text = Handler.gsonHeavy.toJson(new Response("sr", rounds.head))
      broadcast(text)
      Timer.seconds(answerTime + 5) {
        if (rounds.head.getAnswers.getAnswers.isEmpty) {
          startRound()
        } else {
          val answers = Handler.gsonLight.toJson(
            new Response("as", rounds.head.getAnswers))
          broadcast(answers)
          room.startVoting()
          Timer.seconds(voteTime + 1) {
			val kanswers = rounds.head.getAnswers;
            val answers = Handler.gsonLight.toJson(
              new Response("vc", kanswers))
			val winner = room.getPlayer(kanswers.getWinner)
			if(winner!=null) {
				println("winner bonus " + winner.getUsername)
				winner.setTotalVoteCount(winner.getTotalVoteCount + rounds.head.getAcronym.length)
			} else {
				println("no winners")
			}
			val speedBonus = room.getPlayer(kanswers.getSpeeder)
			if(speedBonus!=null) {
				println("speed bonus " + speedBonus.getUsername)
				speedBonus.setTotalVoteCount(speedBonus.getTotalVoteCount + 2)
			} else {
				println("no speed bonus")
			}
            for {
              player <- players
              answer <- Option(rounds.head.getAnswer(player.getUserId))
            } {
              player.setTotalVoteCount(
                player.getTotalVoteCount + answer.getVoteCount
              )
              broadcast(answers)
            }
            Timer.seconds(10) {
              startRound()
            }
          }
        }
      }
    }
  }
}

